### PR TITLE
Minor refactoring of `check_symlink`

### PR DESCRIPTION
**Pull request by muffato**
Hiya,

There are two changes in this pull-request.

The first two commits are a follow up of 10cf953d9c29b282781b0e50c10bd4b3ab7295ef. No need to call `mkdirp` for `symlink_base` since `mkdirp` is anyway called in `create_symlink` to create the parent directory. Then I slightly refactor `check_symlink` to give it a similar interface to `create_symlink`, i.e.: both are only called if the user wants symlinks, and neither need to check the symlink parameter.

The last commit is about stripping `module.tcl` from being shown by `module avail` in the symlinks tree. This is done by making the symlinks point at the `module.tcl` files directly, rather than their enclosing directories. It makes the symlink tree much shorter and nicer to use.
To achieve that, I had to change the module templates to hardcode the path to the modules, because otherwise it couldn't find the `bin/` of the wrapper scripts.

Example trees:
```
containers/
├── golang
│   └── 1.18-rc
│       └── golang-1.18-rc-sha256:c9be2380bfdc0f4da95e1635e914dbd596bc73b618cdf16894a631eb58945156.sif
└── quay.io
    └── biocontainers
        ├── bwa
        │   └── 0.7.17--h7132678_9
        │       └── quay.io-biocontainers-bwa-0.7.17--h7132678_9-sha256:07822e4293a8c59755b295c448b9541db6c9bdbfdedb010bdbdcc1e1e935370f.sif
        ├── bwa-mem2
        │   └── 2.2.1--hd03093a_2
        │       └── quay.io-biocontainers-bwa-mem2-2.2.1--hd03093a_2-sha256:df046077d0771d622d2889039050a995a71430ed155f0f94f52e944bf2785b2a.sif
        ├── cooler
        │   ├── 0.8.11--pyh3252c3a_0
        │   │   └── quay.io-biocontainers-cooler-0.8.11--pyh3252c3a_0-sha256:c954780cf75fd8fe026de342253982a354b6211258e45f66e18d08af68271138.sif
        │   └── 0.8.6--py_0
        │       └── quay.io-biocontainers-cooler-0.8.6--py_0-sha256:1b0e2e9625d9d0beee5befaf866df016a80f364339fbb396ab8d84878a3a0584.sif
        └── samtools
            └── 1.15--h3843a85_0
                └── quay.io-biocontainers-samtools-1.15--h3843a85_0-sha256:d68e1b5f504dc60eb9f2a02eecbac44a63f144e7d455b3fb1a25323c667ca4c4.sif

modules/
├── golang
│   └── 1.18-rc
│       ├── 99-shpc.sh
│       ├── bin
│       │   ├── go
│       │   └── gofmt
│       └── module.tcl
└── quay.io
    └── biocontainers
        ├── bwa
        │   └── 0.7.17--h7132678_9
        │       ├── 99-shpc.sh
        │       ├── bin
        │       │   └── bwa
        │       └── module.tcl
        ├── bwa-mem2
        │   └── 2.2.1--hd03093a_2
        │       ├── 99-shpc.sh
        │       ├── bin
        │       │   └── bwa-mem2
        │       └── module.tcl
        ├── cooler
        │   ├── 0.8.11--pyh3252c3a_0
        │   │   ├── 99-shpc.sh
        │   │   ├── bin
        │   │   │   └── cooler
        │   │   └── module.tcl
        │   └── 0.8.6--py_0
        │       ├── 99-shpc.sh
        │       ├── bin
        │       │   └── cooler
        │       └── module.tcl
        └── samtools
            └── 1.15--h3843a85_0
                ├── 99-shpc.sh
                ├── bin
                │   ├── bgzip
                │   ├── htsfile
                │   ├── samtools
                │   └── tabix
                └── module.tcl

symlinks/
├── bwa
│   └── 0.7.17--h7132678_9 -> /lustre/scratch123/tol/teams/tolit/users/mm49/shpc/singularity-hpc/modules/quay.io/biocontainers/bwa/0.7.17--h7132678_9/module.tcl
├── bwa-mem2
│   └── 2.2.1--hd03093a_2 -> /lustre/scratch123/tol/teams/tolit/users/mm49/shpc/singularity-hpc/modules/quay.io/biocontainers/bwa-mem2/2.2.1--hd03093a_2/module.tcl
├── cooler
│   ├── 0.8.11--pyh3252c3a_0 -> /lustre/scratch123/tol/teams/tolit/users/mm49/shpc/singularity-hpc/modules/quay.io/biocontainers/cooler/0.8.11--pyh3252c3a_0/module.tcl
│   └── 0.8.6--py_0 -> /lustre/scratch123/tol/teams/tolit/users/mm49/shpc/singularity-hpc/modules/quay.io/biocontainers/cooler/0.8.6--py_0/module.tcl
├── golang
│   └── 1.18-rc -> /lustre/scratch123/tol/teams/tolit/users/mm49/shpc/singularity-hpc/modules/golang/1.18-rc/module.tcl
└── samtools
    └── 1.15--h3843a85_0 -> /lustre/scratch123/tol/teams/tolit/users/mm49/shpc/singularity-hpc/modules/quay.io/biocontainers/samtools/1.15--h3843a85_0/module.tcl
```

Example `module avail` output
```
----------------------------------------------- /nfs/users/nfs_m/mm49/nfs/scratch123/shpc/singularity-hpc/modules -----------------------------------------------
golang/1.18-rc/module.tcl                                    quay.io/biocontainers/cooler/0.8.6--py_0/module.tcl           
quay.io/biocontainers/bwa-mem2/2.2.1--hd03093a_2/module.tcl  quay.io/biocontainers/cooler/0.8.11--pyh3252c3a_0/module.tcl  
quay.io/biocontainers/bwa/0.7.17--h7132678_9/module.tcl      quay.io/biocontainers/samtools/1.15--h3843a85_0/module.tcl    

---------------------------------------------- /nfs/users/nfs_m/mm49/nfs/scratch123/shpc/singularity-hpc/symlinks -----------------------------------------------
bwa-mem2/2.2.1--hd03093a_2  bwa/0.7.17--h7132678_9  cooler/0.8.6--py_0  cooler/0.8.11--pyh3252c3a_0  golang/1.18-rc  samtools/1.15--h3843a85_0  
```

Best,
Matthieu

### CODE REVIEW

**Comment by vsoch** [comment]
The first part of these changes look good! For:

> To achieve that, I had to change the module templates to hardcode the path to the modules, because otherwise it couldn't find the bin/ of the wrapper scripts.

This is something @marcodelapierre and I just worked on to remove, meaning that the module installs were portable, and I'm hoping you can find a solution that can preserve that. If not, we can integrate the other changes and continue discussion on how to do that (for some future PR).

**Comment by muffato** [comment]
OK, I see. I could find the comments in the PRs. I'll see what I can do.

**Comment by marcodelapierre** [comment]
happy to help with the module path in the templates

**Comment by vsoch** [comment]
So my understanding is that when we make a symlink tree:

```bash
$ shpc install --symlink-tree python
$ shpc use ./symlinks
```
To produce:
```
[root@528305e80a97 code]# tree symlinks/
symlinks/
`-- python
    `-- 3.9.10 -> /home/vanessa/Desktop/Code/shpc/modules/python/3.9.10

1 directory, 1 file
```
And then do module avail:

```
-------------------------------------------------------- symlinks --------------------------------------------------------
   python/3.9.10/module (D)

```
I think @muffato issue is that it shows module at the end? Or that the tree is slightly deeper than it could be otherwise? I guess I'm not fully understanding the problem or the need for a fix. If it's just an aesthetic thing but the fix takes a step backwards to remove the portability of the install I don't think it's a good direction to go in.

@muffato my suggestion for now, since you did state there are _two_ changes here, is to:

1. remove the hard coded module dir to leave your first set of fixes, and merge these fixes into #502 
2. Review and merge both #502 and #506 since the latter has a bugfix (thank you!)
3. Work on this additional issue in a fresh issue/discussion/pull request.

There are only three of us at the moment so we need to be organized and scoped with respect to how we review and make changes! I'm in full support of all discussion and changes but one thing at a time so we don't get overwhelmed :) Since we already have the symlink tree, and several bug fixes underway let's finish those up first! And let me know if I am misunderstanding and there is a bug/flaw with the current approach that must be addressed.

Let me know what you think!

**Comment by muffato** [comment]
It's mostly a cosmetic change as the module can still be loaded as per usual

Without my change, `module avail` print this:
```
bwa-mem2/2.2.1--hd03093a_2/module.tcl  cooler/0.8.6--py_0/module.tcl           golang/1.18-rc/module.tcl             
bwa/0.7.17--h7132678_9/module.tcl      cooler/0.8.11--pyh3252c3a_0/module.tcl  samtools/1.15--h3843a85_0/module.tcl  
```

With the change, `module.tcl` is stripped out:
```
bwa-mem2/2.2.1--hd03093a_2  bwa/0.7.17--h7132678_9  cooler/0.8.6--py_0  cooler/0.8.11--pyh3252c3a_0  golang/1.18-rc  samtools/1.15--h3843a85_0  
```

I agree with your plan, @vsoch

**Comment by vsoch** [comment]
Sounds good! Ping me if you want any help, want to discuss things, or really whatever I'm here to help :)

**Comment by muffato** [comment]
Done: this PR has now been restricted to the first two commits only

File: shpc/main/modules/__init__.py [hunk 1]
```diff
@@ -207,18 +207,11 @@ def create_symlink(self, module_dir):
             if not os.path.exists(version_file):
                 Path(version_file).touch()
 
-    def check_symlink(self, module_dir, symlink=False):
+    def check_symlink(self, module_dir):
         """
-        Given an install command, if --symblink is provided make sure we have
-        a symlink_base defined in settings and the directory exists.
+        Given an install command, if --symlink-tree is provided make
+        sure we don't already have this symlink in.
```
Comment (vsoch) [inline shpc/main/modules/__init__.py:213]:
```suggestion
        sure we don't already have this symlink in the tree.
```

**Review by vsoch — APPROVED** [review]

**Comment by vsoch** [comment]
Awesome work @muffato ! let's get this merged into the symlink PR branch, and I suspect @marcodelapierre will take a look sometime in the coming weeks, and please open an issue if you want to discuss the tweaks to the symlink.

**Comment by marcodelapierre** [comment]
Hi @vsoch , @muffato , still working in limited capacity these weeks, so getting to these updates...slowly :)
I do have a comment on this one, but as I see it's now merged, I will comment in the appropriate PR

**Comment by vsoch** [comment]
@marcodelapierre please no rush take whatever time you need! We'll be around :)

**Pull request merged**